cmake/CMakeLists.txt

```diff
@@ -233,6 +235,28 @@ if(onnxruntime_DISABLE_RTTI)
   endif()
 endif()
 
+# If this is only enabled in an onnxruntime_ORT_MODEL_FORMAT_ONLY build we don't need ONNX changes
+# as we (currently) only pull in data_type_utils.cc/h which doesn't throw
+if (onnxruntime_NO_EXCEPTIONS)
+  add_compile_definitions("ORT_NO_EXCEPTIONS")
+  add_compile_definitions("MLAS_NO_EXCEPTION")
+
+  if(MSVC)
+    string(REGEX REPLACE "/EHsc" "/EHs-c-" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+    # Eigen throw_std_bad_alloc calls 'new' instead of throwing which results in a nodiscard warning.
+    # It also has unreachable code as there's no good way to avoid EIGEN_EXCEPTIONS being set in macros.h
+    # TODO: see if we can limit the code this is disabled for.
+    string(APPEND CMAKE_CXX_FLAGS " /wd4834 /wd4702")
+    add_compile_definitions("_HAS_EXCEPTIONS=0")
+  else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-exceptions")
+  endif()
+
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGSL_TERMINATE_ON_CONTRACT_VIOLATION")
```

Comment:
Maybe limit setting this GSL define to the debug cxx flags given we disable the contract checking in release builds so (and the intent isn't clear if you GSL_TERMINATE_ON_CONTRACT_VIOLATION and GSL_UNENFORCED_ON_CONTRACT_VIOLATION in those builds)

Reply:
Moved GSL setting back together

---
In reply to: [479834950](https://github.com/microsoft/onnxruntime/pull/4975#discussion_r479834950) [](ancestors = 479834950)